lib/recursive-methods.rb

```diff
@@ -1,49 +1,91 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n), err, does the stack take up a lot of space? theres n numbers of calls
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError, "cant be negative number" if 0 > n
+  if n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n) where n is the length of the string
```

Comment:
Since `s[0...s.length-1]` creates a new string, this algorithm is actually O(n^2)